### ISSUE TITLE
Sharphound.ps1 unhandled exception

### ISSUE THREAD

**Issue by ionrover2** [BloodHoundAD/BloodHound#295]
at commit 848854fad

I am encountering an unhandled exception that is forcing my powershell process to terminate. 

the previous version of sharphound.ps1 works perfectly. I have been using this tool to audit and secure my environment. 

I am using simply `invoke-bloodhound -collectionmethod all`.

I don't have the tools on the computer in my environment to debug the base64 c# binary to find exactly what is throwing the exception. suffice it to say my environment is HUGE and i'm scanning an extremely large number of objects

I am able to gather user data, but it throws while gathering computer data. any advice is appriciated. 

I could not find any powershell logs that would indicate why the process was terminating. AV does not target powershell on the machine i'm running this on.

**Comment by kafetz** [comment]
Experienced the same issue, run it two or three times.
Environment is about 4,500 hosts, running straight onto a DC (Windows 2012 r2).
The program appears to keep running but will never finish (left it for 6 hours and still finish).
The option of the error is 'Close Powershell' and 'Report to Microsoft and close powershell'.
When I quit, it dumps partial json files into a user folder, however, these are unparsable and not in a zip. ``

**Comment by nikallass** [comment]
Same for me.

Was showing this exception:
`Unhandled Exception: System.AggregateException: One or more errors occurred. ---> System.AggregateException: One or more errors occurred. ---> System.AggregateException: One or more errors occurred. ---> System.DirectoryServices.Protocols.LdapException: The supplied credential is invalid.
   at System.DirectoryServices.Protocols.LdapConnection.BindHelper(NetworkCredential newCredential, Boolean needSetCredential)
   at System.DirectoryServices.Protocols.LdapConnection.SendRequestHelper(DirectoryRequest request, Int32& messageID)
   at System.DirectoryServices.Protocols.LdapConnection.BeginSendRequest(DirectoryRequest request, TimeSpan requestTimeout, PartialResultProcessing partialMode, AsyncCallback callback, Object state)
   at SharpHound3.DirectorySearch.<LookupUserInGC>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at SharpHound3.Tasks.NetSessionTasks.<GetNetSessions>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at SharpHound3.Tasks.NetSessionTasks.<ProcessNetSessions>d__0.MoveNext()
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at SharpHound3.SharpHound.<Main>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at SharpHound3.SharpHound.<Main>(String[] args)
`

Fixed it adding catch after line 77:
![image](https://user-images.githubusercontent.com/10115306/74460630-389f8880-4e9e-11ea-98f4-8991fd13735f.png)

**Comment by ionrover2** [comment]
did you or could you create a pull request for this? @nikallass

**Comment by rvazarkar** [comment]
I'm curious what the underlying problem here is. Essentially LDAP is telling you you don't have a valid credential to connect to AD. But if you're in a domain context, this shouldn't really be the case.

If you're looking for more detailed errors, you should use the debug build [here](https://github.com/BloodHoundAD/BloodHound/tree/master/Ingestors/DebugBuilds)

**Comment by nikallass** [comment]
> I'm curious what the underlying problem here is. Essentially LDAP is telling you you don't have a valid credential to connect to AD. But if you're in a domain context, this shouldn't really be the case.
> 
> If you're looking for more detailed errors, you should use the debug build [here](https://github.com/BloodHoundAD/BloodHound/tree/master/Ingestors/DebugBuilds)

That was very strange. I'm using bloodhound in a huge environment (>30000 hosts) for  long time. Before version 3.0 everything was OK. Yesterday I tried new release. I execute SharpHound.exe with all collectors. SharpHound successfully collected all Users.json, collected part of Computers.json and crashed. It is diffcult for me to debug on production environment. I thought maybe I dont have rights for some computer objects and just put down catch statement and it worked. If it is important for you to debug this problem then I could try to run with debug build one more time.

**Comment by nikallass** [comment]
And yes, I was in a domain context (domain user on computer in the same domain).

**Comment by nikallass** [comment]
> did you or could you create a pull request for this? @nikallass

@ionrover2, yes, please:
https://github.com/BloodHoundAD/SharpHound3/pull/1

**Comment by nikallass** [comment]
> I'm curious what the underlying problem here is. Essentially LDAP is telling you you don't have a valid credential to connect to AD. But if you're in a domain context, this shouldn't really be the case.
> 
> If you're looking for more detailed errors, you should use the debug build [here](https://github.com/BloodHoundAD/BloodHound/tree/master/Ingestors/DebugBuilds)

I've tried the latest SharpHound.exe debug build (without my fix) and it crashes the same way:

`PS C:\Temp> .\SharpHound_debug_build.exe --NoSaveCache --CollectionMethod All,Container,GPOLocalGroup -d DOM1.COMPANY.COM -v
----------------------------------------------
Initializing SharpHound at 12:23 on 14.02.2020
----------------------------------------------

Resolved Collection Methods: Group, Sessions, LoggedOn, Trusts, ACL, ObjectProps, LocalGroups, SPNTargets, Container, GPOLocalGroup

[+] Creating Schema map for domain DOM1.COMPANY.COM using path CN=Schema,CN=Configuration,DC=DOM1,DC=COMPANY,DC=COM
[+] Cache File not Found: 0 Objects in cache

[+] Pre-populating Domain Controller SIDS
Status: 0 objects finished (+0) -- Using 21 MB RAM
Status: 8895 objects finished (+8895 296,5)/s -- Using 116 MB RAM
Status: 26552 objects finished (+17651 442,5333)/s -- Using 81 MB RAM
Status: 49118 objects finished (+22566 545,7556)/s -- Using 100 MB RAM
[+] Creating Schema map for domain DOM2 using path CN=Schema,CN=Configuration,DC=DOM2
[+] Creating Schema map for domain DOM2 using path CN=Schema,CN=Configuration,DC=DOM2
Status: 54379 objects finished (+5261 453,1583)/s -- Using 196 MB RAM
Status: 56919 objects finished (+2540 379,46)/s -- Using 221 MB RAM
[+] Creating Schema map for domain WORKGROUP using path CN=Schema,CN=Configuration,DC=WORKGROUP
[+] Creating Schema map for domain DOM1.COMPANY.COM@DOM1.COMPANY.COM using path CN=Schema,CN=Configuration,DC=DOM1,DC=COMPANY,DC=COM@DOM1,DC=COMPANY,DC=COM
Status: 58664 objects finished (+1745 325,9111)/s -- Using 222 MB RAM

Unhandled Exception: System.AggregateException: One or more errors occurred. ---> System.AggregateException: One or more errors occurred. ---> System.AggregateException: One or more errors occurred. ---> System.DirectoryServices.Protocols.LdapException: The supplied credential is invalid.
   at System.DirectoryServices.Protocols.LdapConnection.BindHelper(NetworkCredential newCredential, Boolean needSetCredential)
   at System.DirectoryServices.Protocols.LdapConnection.SendRequestHelper(DirectoryRequest request, Int32& messageID)
   at System.DirectoryServices.Protocols.LdapConnection.BeginSendRequest(DirectoryRequest request, TimeSpan requestTimeout, PartialResultProcessing partialMode, AsyncCallback callback, Object state)
   at SharpHound3.DirectorySearch.<LookupUserInGC>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at SharpHound3.Tasks.NetSessionTasks.<GetNetSessions>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at SharpHound3.Tasks.NetSessionTasks.<ProcessNetSessions>d__0.MoveNext()
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at SharpHound3.SharpHound.<Main>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at SharpHound3.SharpHound.<Main>(String[] args)
`
Created Files:
![image](https://user-images.githubusercontent.com/10115306/74519974-c5425900-4f27-11ea-880c-bb01a8fe5d6d.png)

No more debug logs were provided. No new information.
